workflow/controller/pod/controller.go

```diff
@@ -0,0 +1,322 @@
+			action = removeFinalizer
+		}
+		switch {
+		case podGC.Strategy == wfv1.PodGCOnWorkflowCompletion:
+		case podGC.Strategy == wfv1.PodGCOnPodCompletion:
+		case podGC.Strategy == wfv1.PodGCOnPodSuccess && pod.Status.Phase == apiv1.PodSucceeded:
+			action = deletePod
+		}
+	}
+	if action != noAction {
+		// The workflow is gone, we have no idea when that happened, so lets base around pod transiution
+		lastTransition := podLastTransition(pod)
+		// GetDeleteDelayDuration returns -1 if no duration, we don't care about failure to parse otherwise here
+		delay := time.Duration(0)
+		delayDuration, _ := podGC.GetDeleteDelayDuration()
+		// In the case of a raw delete make sure we've had some time to process it if there was a finalizer
+		if delayDuration < minimumDelay {
+			delayDuration = minimumDelay
+		}
+		if !lastTransition.IsZero() && delayDuration > 0 {
+			delay = time.Until(lastTransition.Add(delayDuration))
+		}
+		c.log.WithFields(logrus.Fields{"action": action, "namespace": pod.Namespace, "podName": pod.Name, "podGC": podGC}).Info("queuing pod", "delay", delay)
+		switch {
+		case delay > 0:
+			c.queuePodForCleanupAfter(pod.Namespace, pod.Name, action, delay)
+		default:
+			c.queuePodForCleanup(pod.Namespace, pod.Name, action)
+		}
+	}
+}
+
+func (c *Controller) addPodEvent(pod *v1.Pod) {
+	c.log.WithField("pod", pod.Name).Info("add pod event")
+	err := c.callBack(pod)
+	if err != nil {
+		c.log.WithField("pod", pod.Name).Warn("callback for pod add failed")
+	}
+	c.commonPodEvent(pod, false)
+}
+
+func (c *Controller) updatePodEvent(old *v1.Pod, new *v1.Pod) {
+	// This is only called for actual updates, where there are "significant changes"
+	c.log.WithField("pod", old.Name).Info("update pod event")
+	err := c.callBack(new)
+	if err != nil {
+		c.log.WithField("pod", new.Name).Warn("callback for pod update failed")
+	}
+	c.commonPodEvent(new, false)
+}
+
+// func (c *Controller) deletePodEvent(pod *v1.Pod) {
+func (c *Controller) deletePodEvent(obj interface{}) {
+	pod, err := podFromObj(obj)
+	if err != nil {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			c.log.Info("error obtaining pod object from tombstone")
+			return
+		}
+		pod, ok = tombstone.Obj.(*apiv1.Pod)
+		if !ok {
+			c.log.Warn("deleted pod last known state not a pod")
+			return
+		}
+	}
+	c.log.WithField("pod", pod.Name).Info("delete pod event")
+	// enqueue the workflow for the deleted pod
+	err = c.callBack(pod)
+	if err != nil {
+		c.log.WithField("pod", pod.Name).Warn("callback for pod delete failed")
+	}
+	// Backstop to remove finalizer if it hasn't already happened, our last chance
+	c.commonPodEvent(pod, true)
+}
+
+func newWorkflowPodWatch(ctx context.Context, clientSet kubernetes.Interface, instanceID, namespace *string) *cache.ListWatch {
+	c := clientSet.CoreV1().Pods(*namespace)
+	// completed=false
+	labelSelector := labels.NewSelector().
+		Add(*workflowReq).
+		// not sure if we should do this
+		Add(*incompleteReq).
```

Comment:
I think we should remove this personally, the controllers job is to manage all pods created by argo.
People upgrading into this version might have trouble with pods not being handled with this remaining the same.

Reply:
There is no path whereby the pods have the finalizer and are not `label`led as completed. In this state they drop out of view of the informer, yes. Monitoring pods that are marked as completed is a waste of resources, owner reference to the workflow does all the necessary work.